tachyon/zk/plonk/vanishing/prover_vanishing_argument.h

```diff
@@ -0,0 +1,322 @@
+                                      size_t chunk_size) {
+                        size_t index = chunk_idx * chunk_size;
+                        for (F& h : chunk) {
+                          h *= t_evaluations[index % t_evaluations.size()];
+                          ++index;
+                        }
+                      });
+
+    return ExtendedEvals(std::move(evaluations));
+  }
+
+  // This takes us from the extended evaluation domain and gets us the quotient
+  // polynomial coefficients.
+  //
+  // This function will panic if the provided vector is not the correct length.
+  constexpr ExtendedPoly ExtendedToCoeff(
+      const ExtendedEvals& evals, const ExtendedDomain* extended_domain) const {
+    CHECK_EQ(evals.NumElements(), extended_domain->size());
+
+    ExtendedPoly poly = extended_domain->IFFT(evals);
+
+    // Distribute powers to move from coset; opposite from the
+    // transformation we performed earlier.
+    DistributePowersZeta(poly, false);
+
+    return poly;
+  }
+
+  // Given a |poly| of coefficients  [a₀, a₁, a₂, ...], this returns
+  // [a₀, ζa₁, ζ²a₂, a₃, ζa₄, ζ²a₅, a₆, ...], where ζ is a cube root of unity in
+  // the multiplicative subgroup with order (p - 1), i.e. ζ³ = 1.
+  //
+  // |into_coset| should be set to true when moving into the coset, and false
+  // when moving out. This toggles the choice of ζ.
+  void DistributePowersZeta(ExtendedPoly& poly, bool into_coset) const {
+    // Calculate ζ = g^((2ˢ * T) / 3)
+    CHECK_EQ(F::Config::kTrace % math::BigInt<F::kLimbNums>(3),
+             math::BigInt<F::kLimbNums>(0));
+    F zeta_inv = F::FromMontgomery(F::Config::kSubgroupGenerator)
+                     .Pow(F(2).Pow(F::Config::kTwoAdicity).ToBigInt() *
+                          F::Config::kTrace / math::BigInt<F::kLimbNums>(3));
+    F zeta = zeta_inv.Square();
+    std::vector<F> coset_powers{into_coset ? zeta : zeta_inv,
+                                into_coset ? zeta_inv : zeta};
+
+    // FIXME(TomTaehoonKim): Avoid copy.
+    std::vector<F> coeffs = poly.coefficients().coefficients();
+    base::Parallelize(coeffs,
+                      [&coset_powers](absl::Span<F> chunk, size_t chunk_idx,
+                                      size_t chunk_size) {
+                        size_t i = chunk_idx * chunk_size;
+                        for (F& a : chunk) {
+                          // Distribute powers to move into/from coset
+                          size_t j = i % (coset_powers.size() + 1);
+                          if (j != 0) {
+                            a *= coset_powers[j - 1];
+                          }
+                          ++i;
+                        }
+                      });
+  }
+
+  bool Construct(Prover<PCSTy>* prover,
+                 const ExtendedEvals& linear_combination_of_gates,
+                 size_t quotient_poly_degree,
+                 Constructed<PCSTy>* constructed_out) {
```

Comment:
```suggestion
                 Constructed<PCSTy>* constructed_out) && {
```

Reply:
I don't think this function fits for r-value ref-qualifier.